### PR TITLE
feat: tutorials opt in / opt out

### PR DESCRIPTION
**Pull request by kendallstrautman**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/vault/tutorials) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1201987217085805) 🎟️

## 🗒️ What

This PR adds the opt-out workflow for tutorials. A new `optInOutSlot` was added to the sidebar sidecar layout. That should render the opt out component, which shows a button and handles the dialog as well. 

This preview is linked to [this branch in Learn](https://github.com/hashicorp/learn/pull/4480), so you can click between the two. For release we'll need to update that base url. 

The goal of this work is to get the general opt-out flow in place and then handle any UI refinements later. 

## 📸 Design Screenshots

[See design here. ](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Dev-Portal?node-id=8356%3A92080) They also adjusted things to use a select dropdown for the dialog, [which can be seen here. ](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Dev-Portal?node-id=9090%3A123877)

## 🧪 Testing

- Visit [waypoint](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/waypoint/tutorials) and [vault](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/vault/tutorials) tutorials pages. Verify that no opt out button renders since the opt-in hasn't been set yet. 
- Now visit this [learn preview to opt-in](https://learn-git-ksspike-opt-in-redirects-hashicorp.vercel.app/waypoint). That should redirect you back to this preview. 
- Click around to a tutorial or collection, hit opt out and open the dialog. Opt out to test the full flow. 

## 💭 Anything else?

The UI still needs to be polished a bit with the dialog but I want to get this core functionality in. I haven't styled the select options because I didn't have time and that feels like a _nice to have_. 

The opt out button can look weird on tutorial pages that have really long titles since we haven't implemented breadcrumb truncation yet. I'm not planning on addressing that in this PR. 

We will be [implementing the toast](https://app.asana.com/0/1201903760348480/1201979557016120) in a follow up PR. 

The form data doesn't submit to anywhere yet, that is follow up work.

### CODE REVIEW

File: src/components/button/index.tsx [hunk 1]
```diff
@@ -18,6 +18,7 @@ const Button = ({
   size = 'medium',
   text,
   type = 'button',
+  form,
```
Comment (kendallstrautman) [inline src/components/button/index.tsx:21]: 🤔  We could consider spreading props here so we don't need to add native button attributes to this component

File: src/components/dialog/index.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { DialogOverlay, DialogContent } from '@reach/dialog'
```
Comment (kendallstrautman) [inline src/components/dialog/index.tsx:1]: I chose to use [reach dialog](https://reach.tech/dialog/) here to go with previous convention. This is a wrapper for the dialog that sets custom base styles.

File: src/components/opt-in-out/helpers/get-learn-redirect-path.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { ProductOption } from 'lib/learn-client/types'
+import { PLATFORM_OPTIONS } from '..'
+
+export function getLearnRedirectPath(
+  currentPath: string,
+  product?: ProductOption
+) {
+  const originalUrl = new URL(currentPath, PLATFORM_OPTIONS['learn'].base_url)
+  const finalUrl = new URL(PLATFORM_OPTIONS['learn'].base_url)
```
Comment (kendallstrautman) [inline src/components/opt-in-out/helpers/get-learn-redirect-path.ts:9]: Redirecting back to learn home page, the middleware on learn will handle the final redirect based on the `path`

File: src/components/opt-in-out/index.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  'vault-io': {
+    base_url: 'https://www.vaultproject.io/',
+    getRedirectPath: function TODO() {
+      return 'test'
+    },
+    cookieKey: 'vault-io-beta-opt-in',
+  },
+}
+
+// @TODO - check query, if 'optInFrom' is there, fire welcome toast
```
Comment (kendallstrautman) [inline src/components/opt-in-out/index.tsx:35]: Will handle toast in follow up work

File: src/components/opt-in-out/index.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    return null
+  }
+
+  const openDialog = () => setShowDialog(true)
+  const closeDialog = () => setShowDialog(false)
+  const url =
+    redirectPath || PLATFORM_OPTIONS[platform].getRedirectPath(router.asPath)
+
+  function handleOptOut() {
+    // @TODO - handle form submit
```
Comment (kendallstrautman) [inline src/components/opt-in-out/index.tsx:53]: We'll handle form submit when we know where the feedback data will live.

File: src/layouts/sidebar-sidecar/index.tsx [hunk 6]
```diff
@@ -96,7 +97,10 @@ const SidebarSidecarLayoutContent = ({
         <div className={s.mainArea}>
           <div className={s.main}>
             <main id="main">
-              {breadcrumbLinks && <BreadcrumbBar links={breadcrumbLinks} />}
+              <span className={s.breadcrumbOptOutGroup}>
+                {breadcrumbLinks && <BreadcrumbBar links={breadcrumbLinks} />}
+                {optInOutSlot && optInOutSlot}
```
Comment (kendallstrautman) [inline src/layouts/sidebar-sidecar/index.tsx:102]: Adding this slot here for the opt-in component on the layout.

File: src/pages/_middleware.ts [hunk 7]
```diff
@@ -2,6 +2,9 @@ import { NextResponse } from 'next/server'
 import type { NextFetchEvent, NextRequest } from 'next/server'
 import redirects from 'data/_redirects.generated.json'
 
+export type OptInPlatformOption = 'vault-io' | 'waypoint-io' | 'learn'
+const OPT_IN_MAX_AGE = 60 * 60 * 24 * 90 //90 days
```
Comment (kendallstrautman) [inline src/pages/_middleware.ts:6]: Anyone have suggestions on this lenght?

File: src/pages/_middleware.ts [hunk 8]
```diff
@@ -72,6 +75,17 @@ export function middleware(req: NextRequest, ev: NextFetchEvent) {
     return NextResponse.redirect(url, permanent ? 308 : 307)
   }
 
+  // Handle Opt-in cookies
+  const response = NextResponse.next()
```
Comment (kendallstrautman) [inline src/pages/_middleware.ts:79]: TODO: add a note on this file explaining the purpose of this cookie.

File: src/components/opt-in-out/index.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import { PlatformOptionsType, OptInOutProps } from './types'
+
+// Could these go in the config? or I could source the base urls elsewhere
+export const PLATFORM_OPTIONS: PlatformOptionsType = {
+  learn: {
+    base_url: 'https://learn-git-ksspike-opt-in-redirects-hashicorp.vercel.app', // FOR TESTING PURPOSES NEED TO UPDATE for - 'https://learn.hashicorp.com/'
+    getRedirectPath: getLearnRedirectPath,
+    cookieKey: 'learn-beta-opt-in',
+  },
+  'waypoint-io': {
```
Comment (zchsh) [inline src/components/opt-in-out/index.tsx:19]: ⛏️ Nitpick thought - Should these keys match the `platformOptions` enum in `opt-out-form`?

Or maybe `opt-out-form`'s `platformOptions` enum should match these keys? (slash should match `OptInPlatformOption` from the `_middleware` file)

**Review by zchsh — APPROVED** [review]
✅  Ran through the `Testing` steps here, and did a relatively thorough review of the code, and everything looks good to me 🙇

I found a few minor nitpicks, but nothing felt blocking. Eg. the enum name nitpicks seem more dot-io than learn, and this PR seems focused on learn, so definitely on board with deferring if we want.

Seems like a great system 💯

File: src/pages/_middleware.ts (hunk 7 above)
Comment (kendallstrautman) [reply src/pages/_middleware.ts:6]: That makes sense! It just sounded v long when I was implementing but I'll update. I think you're right, might as well assume folks can opt in until GA.

File: src/components/dialog/index.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  children: ReactNode
+  label: string
+}
+
+const overlayStyles = {
+  background: 'var(--token-color-border-strong',
+  zIndex: 100,
+}
+
+const dialogStyles = {
```
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/dialog/index.tsx:16]: Ah okay, I think that'd be sweet. Also I think @alexcarpenter made a dialog component as well? So we should make a follow up PR to unify the two.

File: src/components/opt-in-out/components/opt-out-form/index.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
+import { IconX16 } from '@hashicorp/flight-icons/svg-react/x-16'
+import Button from 'components/button'
+import Text from 'components/text'
+import s from './opt-out-form.module.css'
+
+interface OptOutFormProps {
+  onSubmit(): void
+  onDismiss(): void
+  platform: 'vaul-io' | 'waypoint-io' | 'learn'
```
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/opt-in-out/components/opt-out-form/index.tsx:9]: I like that, also I just saw a typo in my original type ha

**Review by BrandonRomano — APPROVED** [review]

File: src/components/dialog/index.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (alexcarpenter) [reply src/components/dialog/index.tsx:16]: Here is this dialog component I had made for the feedback demo https://github.com/hashicorp/dev-portal/tree/ac.feedback-demo/src/components/dialog

**Pull request merged**